Index: git/sysdeps/unix/sysv/linux/timer_routines.c
===================================================================
--- git.orig/sysdeps/unix/sysv/linux/timer_routines.c
+++ git/sysdeps/unix/sysv/linux/timer_routines.c
@@ -157,8 +157,9 @@ __start_helper_thread (void)
      and should go away automatically when canceled.  */
   pthread_attr_t attr;
   (void) pthread_attr_init (&attr);
-  (void) pthread_attr_setstacksize (&attr, __pthread_get_minstack (&attr));
+  //(void) pthread_attr_setstacksize (&attr, __pthread_get_minstack (&attr));

+  (void) pthread_attr_setstacksize (&attr, ( 2 * 1024 * 1024));
   /* Block all signals in the helper thread but SIGSETXID.  To do this
      thoroughly we temporarily have to block all signals here.  The
      helper can lose wakeups if SIGCANCEL is not blocked throughout,
