Date: Feb 28 00:31:58 2019

From: aef8e592533d1d806a7b46ce836c6929bd9569d1 Apr 01 18:05:22 2019

Subject: [PATCH] RDK-22987: enforcing TLS 1.2

	TLS 1.2 minimal

Source: COMCAST

License: LGPL-2.1-or-later

Upstream-Status: Pending

Signed-off-by: Igor Borovkov <Igor_Borovkov@cable.comcast.com>	
---

Index: glib-networking-2.72.0/tls/openssl/gtlsclientconnection-openssl.c
===================================================================
--- glib-networking-2.72.0.orig/tls/openssl/gtlsclientconnection-openssl.c
+++ glib-networking-2.72.0/tls/openssl/gtlsclientconnection-openssl.c
@@ -468,6 +468,7 @@ g_tls_client_connection_openssl_initable
     return FALSE;
 
   /* Only TLS 1.2 or higher */
+  if (!g_getenv("NO_FORCE_TLS_1_2")) {
   options = SSL_OP_NO_TICKET |
             SSL_OP_NO_COMPRESSION |
 #ifdef SSL_OP_NO_TLSv1_1
@@ -476,6 +477,9 @@ g_tls_client_connection_openssl_initable
             SSL_OP_NO_SSLv2 |
             SSL_OP_NO_SSLv3 |
             SSL_OP_NO_TLSv1;
+  } else {
+  options = SSL_OP_NO_TICKET;
+  }
   SSL_CTX_set_options (client->ssl_ctx, options);
 
   SSL_CTX_clear_options (client->ssl_ctx, SSL_OP_LEGACY_SERVER_CONNECT);
