diff --git a/tls/openssl/gtlsconnection-openssl.c b/tls/openssl/gtlsconnection-openssl.c
index 06356c5..6d5deac 100644
--- a/tls/openssl/gtlsconnection-openssl.c
+++ b/tls/openssl/gtlsconnection-openssl.c
@@ -135,7 +135,16 @@ end_openssl_io (GTlsConnectionOpenssl  *openssl,
     }

   if (err_code == SSL_ERROR_ZERO_RETURN)
-    return G_TLS_CONNECTION_BASE_OK;
+    {
+      if (SSL_get_shutdown(ssl) != 0)
+        {
+          g_set_error_literal (error, G_IO_ERROR,
+                                     G_IO_ERROR_CONNECTION_CLOSED,
+                                     _("Connection terminated unexpectedly"));
+          return G_TLS_CONNECTION_BASE_ERROR;
+        }
+      return G_TLS_CONNECTION_BASE_OK;
+    }

   if (status == G_TLS_CONNECTION_BASE_OK ||
       status == G_TLS_CONNECTION_BASE_WOULD_BLOCK ||
