Index: openssl-1.1.1g/crypto/ocsp/ocsp_prn.c
===================================================================
--- openssl-1.1.1g.orig/crypto/ocsp/ocsp_prn.c
+++ openssl-1.1.1g/crypto/ocsp/ocsp_prn.c
@@ -14,6 +14,8 @@
 #include "internal/cryptlib.h"
 #include <openssl/pem.h>

+#define CUSTOM_LOG "/opt/logs/netstat.log"
+
 static int ocsp_certid_print(BIO *bp, OCSP_CERTID *a, int indent)
 {
     BIO_printf(bp, "%*sCertificate ID:\n", indent, "");
@@ -61,11 +63,14 @@ const char *OCSP_response_status_str(lon

 const char *OCSP_cert_status_str(long s)
 {
+    BIO *outbio = BIO_new_file(CUSTOM_LOG, "a");
     static const OCSP_TBLSTR cstat_tbl[] = {
         {V_OCSP_CERTSTATUS_GOOD, "good"},
         {V_OCSP_CERTSTATUS_REVOKED, "revoked"},
         {V_OCSP_CERTSTATUS_UNKNOWN, "unknown"}
     };
+    BIO_printf(outbio, "cert_status: %s\n", table2string(s, cstat_tbl));
+    BIO_free(outbio);
     return table2string(s, cstat_tbl);
 }

Index: openssl-1.1.1g/include/openssl/ssl.h
===================================================================
--- openssl-1.1.1g.orig/include/openssl/ssl.h
+++ openssl-1.1.1g/include/openssl/ssl.h
@@ -1680,6 +1680,8 @@ __owur X509 *SSL_get_peer_certificate(co

 __owur STACK_OF(X509) *SSL_get_peer_cert_chain(const SSL *s);

+int SSL_canary_cert_chain_logging(STACK_OF(X509) *r);
+
 __owur int SSL_CTX_get_verify_mode(const SSL_CTX *ctx);
 __owur int SSL_CTX_get_verify_depth(const SSL_CTX *ctx);
 __owur SSL_verify_cb SSL_CTX_get_verify_callback(const SSL_CTX *ctx);
Index: openssl-1.1.1g/ssl/build.info
===================================================================
--- openssl-1.1.1g.orig/ssl/build.info
+++ openssl-1.1.1g/ssl/build.info
@@ -7,7 +7,7 @@ SOURCE[../libssl]=\
         methods.c   t1_lib.c  t1_enc.c tls13_enc.c \
         d1_lib.c  record/rec_layer_d1.c d1_msg.c \
         statem/statem_dtls.c d1_srtp.c \
-        ssl_lib.c ssl_cert.c ssl_sess.c \
+        ssl_lib.c ssl_canary.c ssl_cert.c ssl_sess.c \
         ssl_ciph.c ssl_stat.c ssl_rsa.c \
         ssl_asn1.c ssl_txt.c ssl_init.c ssl_conf.c  ssl_mcnf.c \
         bio_ssl.c ssl_err.c tls_srp.c t1_trce.c ssl_utst.c \
Index: openssl-1.1.1g/ssl/ssl_canary.c
===================================================================
--- /dev/null
+++ openssl-1.1.1g/ssl/ssl_canary.c
@@ -0,0 +1,48 @@
+#include <stdio.h>
+#include "ssl_local.h"
+
+#define CUSTOM_LOG "/opt/logs/netstat.log"
+
+int SSL_canary_cert_chain_logging(STACK_OF(X509) *r) {
+    char buf[BUFSIZ], *str = NULL;
+    int i = 0, j = 0;
+    BIO *outbio = BIO_new_file(CUSTOM_LOG, "a");
+    if (r != NULL) {
+        for (i = 0; i < sk_X509_num(r); i++) {
+            if (i == 0) {
+                X509_NAME *subj = X509_get_subject_name(sk_X509_value(r, i));
+                for (j = 0; j < X509_NAME_entry_count(subj); j++) {
+                    X509_NAME_ENTRY *e = X509_NAME_get_entry(subj, j);
+                    ASN1_STRING *d = X509_NAME_ENTRY_get_data(e);
+                    str = ASN1_STRING_data(d);
+                    if(str != NULL) {
+                        if(j == X509_NAME_entry_count(subj) - 1) {
+                            BIO_printf(outbio, "\nEnd point connection :%s\n", str);
+                        }
+                    }
+                }
+            }
+
+            X509_NAME_oneline(X509_get_subject_name(sk_X509_value(r, i)),
+                    buf, sizeof(buf));
+            BIO_printf(outbio, "%2d s:%s\n", i, buf);
+            X509_NAME_oneline(X509_get_issuer_name(sk_X509_value(r, i)),
+                    buf, sizeof(buf));
+            BIO_printf(outbio, "   i:%s\n", buf);
+            BIO_printf(outbio, "+++++++certificate validity++++++\n");
+            BIO_printf(outbio, " notBefore: ");
+            ASN1_TIME_print(outbio, X509_get_notBefore(sk_X509_value(r, i)));
+            BIO_printf(outbio, "\n notAfter : ");
+            ASN1_TIME_print(outbio, X509_get_notAfter(sk_X509_value(r, i)));
+            if(str != NULL) {
+               BIO_printf(outbio, "   End_point :%s\n", str);
+            }
+            if (X509_cmp_current_time(X509_get_notAfter(sk_X509_value(r, i))) == -1) {
+                BIO_printf(outbio, "\n Certificate is expired/revoked for End_point :%s\n", str);
+            }
+        }
+    }
+    BIO_printf(outbio, "Length of the Cert chain: %d\n", i);
+    BIO_free(outbio);
+    return 0;
+}
Index: openssl-1.1.1g/ssl/ssl_lib.c
===================================================================
--- openssl-1.1.1g.orig/ssl/ssl_lib.c
+++ openssl-1.1.1g/ssl/ssl_lib.c
@@ -1517,17 +1517,24 @@ int SSL_has_pending(const SSL *s)
 X509 *SSL_get_peer_certificate(const SSL *s)
 {
     X509 *r;
+    STACK_OF(X509) *certchain;

-    if ((s == NULL) || (s->session == NULL))
+    if ((s == NULL) || (s->session == NULL)) {
         r = NULL;
+	certchain = NULL;
+    }
     else
+    {
         r = s->session->peer;
+	certchain = s->session->peer_chain;
+    }

     if (r == NULL)
         return r;

     X509_up_ref(r);

+    SSL_canary_cert_chain_logging(certchain);
     return r;
 }

Index: openssl-1.1.1g/util/libssl.num
===================================================================
--- openssl-1.1.1g.orig/util/libssl.num
+++ openssl-1.1.1g/util/libssl.num
@@ -498,3 +498,4 @@ SSL_CTX_get_recv_max_early_data
 SSL_CTX_set_recv_max_early_data         499	1_1_1	EXIST::FUNCTION:
 SSL_CTX_set_post_handshake_auth         500	1_1_1	EXIST::FUNCTION:
 SSL_get_signature_type_nid              501	1_1_1a	EXIST::FUNCTION:
+SSL_canary_cert_chain_logging           32767	1_1_1g	EXIST::FUNCTION:
