Date: Jan 4, 2021
From: mkrish802 <Manoj_Krishnan@comcast.com>
Source: COMCAST
Subject: Drop root privileges for lighttpd process
Signed-off-by: mkrish802 <Manoj_Krishnan@comcast.com>
Upstream-Status: Pending
Index: ssaman560/BPI-4.2/build-img/tmp/work/cortexa53-rdk-linux/lighttpd/1.4.74/lighttpd-1.4.74/src/server.c
===================================================================
--- ssaman560.orig/BPI-4.2/build-img/tmp/work/cortexa53-rdk-linux/lighttpd/1.4.74/lighttpd-1.4.74/src/server.c
+++ ssaman560/BPI-4.2/build-img/tmp/work/cortexa53-rdk-linux/lighttpd/1.4.74/lighttpd-1.4.74/src/server.c
@@ -18,6 +18,7 @@
 #include "network_write.h"  /* network_write_show_handlers() */
 #include "reqpool.h"        /* request_pool_init() request_pool_free() */
 #include "response.h"       /* http_dispatch[] strftime_cache_reset() */
+#include "cap.h"
 
 #ifdef HAVE_VERSIONSTAMP_H
 # include "versionstamp.h"
@@ -182,6 +183,7 @@ static const struct http_dispatch h1_1_d
   .send_1xx          = h1_send_1xx
 };
 
+static cap_user appcaps;
 static int oneshot_fd = 0;
 static int oneshot_fdout = -1;
 static fdnode *oneshot_fdn = NULL;
@@ -2309,6 +2311,21 @@ static int main_init_once (void) {
     return 1;
 }
 
+void drop_root()
+{
+   appcaps.caps = NULL;
+   appcaps.user_name = NULL;
+   bool ret = false;
+   ret = isBlocklisted();
+   if(!ret)
+   {
+	   init_capability();
+	   drop_root_caps(&appcaps);
+	   update_process_caps(&appcaps);
+	   read_capability(&appcaps);
+   }
+}
+
 #ifndef server_status_running
 #define server_status_running(srv) do { } while (0)
 #endif
