Index: opkg-0.4.2/libopkg/opkg.c
===================================================================
--- opkg-0.4.2.orig/libopkg/opkg.c
+++ opkg-0.4.2/libopkg/opkg.c
@@ -236,6 +236,11 @@ int opkg_install_package(const char *pac
         } else {
             fprintf(LOG, "opkg_install_package: rdm_signature_verify success with return code = %d\n", err);
             system("touch /tmp/.opkg_rdm_sign_verify_success");
+
+            pdata.pkg = pkg_new();
+            pdata.pkg->local_filename = xstrdup(local_filename);
+            progress(&pdata, 100, progress_callback, user_data);
+
             free(rdm_cache_dir);
             free(local_filename);
             fclose(LOG);
