Date: Nov 15, 2023
From: Aravindan NarasimhapuramChakravarthy <aravindan_narasimhapuramchakravarthy@comcast.com>
Source: COMCAST
Subject: Breakpad Changes for gcc11
Signed-off-by: Aravindan NarasimhapuramChakravarthy <aravindan_narasimhapuramchakravarthy@comcast.com>
Index: git/src/common/memory_allocator.h
===================================================================
--- git.orig/src/common/memory_allocator.h
+++ git/src/common/memory_allocator.h
@@ -161,9 +161,12 @@ class PageAllocator {
 
 // Wrapper to use with STL containers
 template <typename T>
-struct PageStdAllocator : public std::allocator<T> {
-  typedef typename std::allocator<T>::pointer pointer;
-  typedef typename std::allocator<T>::size_type size_type;
+struct PageStdAllocator {
+  using AllocatorTraits = std::allocator_traits<std::allocator<T>>;
+  using value_type = typename AllocatorTraits::value_type;
+  using pointer = typename AllocatorTraits::pointer;
+  using difference_type = typename AllocatorTraits::difference_type;
+  using size_type = typename AllocatorTraits::size_type;
 
   explicit PageStdAllocator(PageAllocator& allocator) : allocator_(allocator),
                                                         stackdata_(NULL),

