Date: Jun 12 2024
From: anaras440_comcast <aravindan_narasimhapuramchakravarthy@comcast.com>
Subject: Add NetworkManager component in RDKE
Source: COMCAST
Signed-off-by: anaras440_comcast <aravindan_narasimhapuramchakravarthy@comcast.com>
Index: git/presets/90-systemd.preset
===================================================================
--- git.orig/presets/90-systemd.preset
+++ git/presets/90-systemd.preset
@@ -25,6 +25,9 @@ enable systemd-userdbd.socket
 enable systemd-pstore.service
 enable systemd-boot-update.service
 
+enable NetworkManager-dispatcher.service
+enable dbus-org.freedesktop.nm-dispatcher.service
+
 disable console-getty.service
 disable debug-shell.service
 

