Index: git/src/timesync/timesyncd.c
===================================================================
--- git.orig/src/timesync/timesyncd.c
+++ git/src/timesync/timesyncd.c
@@ -130,6 +130,15 @@ int main(int argc, char *argv[]) {
         if (r < 0)
                 log_warning_errno(r, "Failed to parse configuration file: %m");

+
+
+
+       struct timespec uptime;
+       clock_gettime(CLOCK_MONOTONIC_RAW, &uptime);
+       long long uptime_ms = (long long)uptime.tv_sec * 1000 + uptime.tv_nsec / 1000000;
+
+       log_info("NTP Client process started at %lld ms", uptime_ms);
+
         log_debug("systemd-timesyncd running as pid " PID_FMT, getpid());
         sd_notify(false,
                   "READY=1\n"
