Date: Jun 19, 2024
From: rgangi643 <ramakrishna_gangisetty@comcast.com>
Source: COMCAST
Subject: mDNSResponder branch changed from 1790 to  1556
Signed-off-by: rgangi643 <ramakrishna_gangisetty@comcast.com>
diff --git a/mDNSPosix/PosixDaemon.c b/mDNSPosix/PosixDaemon.c
index 9a0f692..4667127 100644
--- a/mDNSPosix/PosixDaemon.c
+++ b/mDNSPosix/PosixDaemon.c
@@ -1,6 +1,6 @@
 /* -*- Mode: C; tab-width: 4; c-file-style: "bsd"; c-basic-offset: 4; fill-column: 108; indent-tabs-mode: nil; -*-
  *
- * Copyright (c) 2003-2020 Apple Inc. All rights reserved.
+ * Copyright (c) 2003-2024 Apple Inc. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -184,6 +184,9 @@ int main(int argc, char **argv)
 
     ParseCmdLineArgs(argc, argv);
 
+    // Enable mDNSResponder logging by default.
+    mDNS_LoggingEnabled = mDNStrue;
+
     LogMsg("%s starting", mDNSResponderVersionString);
 
     err = mDNS_Init(&mDNSStorage, &PlatformStorage, gRRCache, RR_CACHE_SIZE, mDNS_Init_AdvertiseLocalAddresses,
diff --git a/mDNSShared/PlatformCommon.c b/mDNSShared/PlatformCommon.c
index 1818a78..42770c2 100644
--- a/mDNSShared/PlatformCommon.c
+++ b/mDNSShared/PlatformCommon.c
@@ -1,6 +1,6 @@
 /* -*- Mode: C; tab-width: 4; c-file-style: "bsd"; c-basic-offset: 4; fill-column: 108; indent-tabs-mode: nil; -*-
  *
- * Copyright (c) 2004-2021 Apple Inc. All rights reserved.
+ * Copyright (c) 2004-2024 Apple Inc. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -300,8 +300,9 @@ mDNSexport void mDNSPlatformWriteLogMsg(const char *ident, const char *buffer, m
             case MDNS_LOG_ERROR:     syslog_level = LOG_ERR;     break;
             case MDNS_LOG_WARNING:   syslog_level = LOG_WARNING; break;
             case MDNS_LOG_DEFAULT:   syslog_level = LOG_NOTICE;  break;
-            case MDNS_LOG_INFO:      syslog_level = LOG_INFO;    break;
-            case MDNS_LOG_DEBUG:     syslog_level = LOG_DEBUG;   break;
+            // Enable info/debug logging for mDNSResponder.
+            case MDNS_LOG_INFO:      syslog_level = LOG_NOTICE;  break;
+            case MDNS_LOG_DEBUG:     syslog_level = LOG_NOTICE;  break;
             default:                 syslog_level = LOG_NOTICE;  break;
         }
 
