Date: 19, Dec 2022
Subject: Logging for DAD failure
Source: COMCAST 793ba608a453f9868bf6ad53447a4e9aad26b09c
Signed-off-by:Aadhithan_PE@comcast.com

Index: git/AddrMgr/AddrIA.cpp
===================================================================
--- git.orig/AddrMgr/AddrIA.cpp
+++ git/AddrMgr/AddrIA.cpp
@@ -436,6 +436,7 @@ enum EAddrStatus TAddrIA::getTentative()
 	case ADDRSTATUS_YES:
 	    Log(Warning) << "DAD failed. Address " << ptrAddr->get()->getPlain() 
 			 << " was detected as tentative." << LogEnd;
+        system("echo DAD failed. Address was detected as tentative. >> /rdklogs/logs/Consolelog.txt.0");
 	    this->Tentative = ADDRSTATUS_YES;
 	    return ADDRSTATUS_YES;
 	case ADDRSTATUS_NO:
@@ -458,6 +459,7 @@ enum EAddrStatus TAddrIA::getTentative()
 	    default:
 		Log(Error) << "DAD inconclusive. Unable to dermine " << ptrAddr->get()->getPlain() 
 			   << " address state. Assuming NOT TENTATIVE." << LogEnd;
+        system("echo DAD inconclusive. Unable to dermine address state. Assuming NOT TENTATIVE. >> /rdklogs/logs/Consolelog.txt.0");
 		ptrAddr->setTentative(ADDRSTATUS_NO);
 		break;
             }
