Index: git/sbin/update-ca-certificates
===================================================================
--- git.orig/sbin/update-ca-certificates
+++ git/sbin/update-ca-certificates
@@ -228,7 +228,8 @@ if [ -f "$TEMPBUNDLE" ]
 then
   chmod 0644 "$TEMPBUNDLE"
   mv -f "$TEMPBUNDLE" "$CERTSDIRBUNDLE"
-  ln -sf "${CERTSDIRBUNDLE##$SYSROOT}" "$CERTBUNDLE"
+  CERTSDIRBUNDLE_LN="$(echo ${CERTSDIRBUNDLE} | sed -e "s|^$SYSROOT||" -e "s|^/|$FSROOT|" )"
+  ln -sf "${CERTSDIRBUNDLE_LN}" "$CERTBUNDLE"
   # Restore proper SELinux label after moving the file
   [ -x /sbin/restorecon ] && /sbin/restorecon "$CERTSDIRBUNDLE" >/dev/null 2>&1
 fi
