Date: Aug 28, 2023
From: smuthu545 <Sindhuja_Muthukrishnan@comcast.com>
Source: COMCAST
Subject: 1.Integrate heaptrack and check
2. Check if backtraces are unwinding properly
Signed-off-by: smuthu545 <Sindhuja_Muthukrishnan@comcast.com>
Index: git/src/interpret/heaptrack_interpret.cpp
===================================================================
--- git.orig/src/interpret/heaptrack_interpret.cpp
+++ git/src/interpret/heaptrack_interpret.cpp
@@ -440,6 +440,13 @@ int main(int /*argc*/, char** /*argv*/)
                 if (fileName == "x") {
                     fileName = exe;
                 }
+                const auto dbg_symbols = getenv("DEBUG_SYMBOLS");
+                if (dbg_symbols) {
+                std::string outputFilename = dbg_symbols + fileName.substr(1);
+                size_t lastSlashPos = outputFilename.rfind('/');
+                outputFilename.insert(lastSlashPos, "/.debug");
+                fileName = outputFilename;
+                }
                 const char* internedString = nullptr;
                 const auto moduleIndex = data.intern(fileName, &internedString);
                 uintptr_t addressStart = 0;
