Date: Mon, 6 Oct 2021 
From: Unknown
Subject: CLOCK_REALTIME build error with disable_threads
Source: Comcast
License: Unknown
Upstream-Status: Pending
Signed-off-by: Rense Jacob <rense_jacob2@cable.comcast.com>

Index: bind-9.11.22/lib/dns/dns64.c
===================================================================
--- bind-9.11.22.orig/lib/dns/dns64.c
+++ bind-9.11.22/lib/dns/dns64.c
@@ -14,6 +14,8 @@

 #include <stdbool.h>

+#include <time.h>
+
 #include <isc/list.h>
 #include <isc/mem.h>
 #include <isc/netaddr.h>
 