#########################################################################
Date: Sept 06, 2023 17:30 PM
From:
Subject: auto-accept bluetooth coonection request
Source: Sky
License:
Upstream-Status: Pending
Signed-off-by: abhishek gupta <abhishek.gupta@sky.uk>
#########################################################################

diff --git a/src/device.c b/src/device.c
index 69f98e488a..5f9ad227d9 100644
--- a/src/device.c
+++ b/src/device.c
@@ -6168,12 +6168,23 @@ int device_confirm_passkey(struct btd_device *device, uint8_t type,
 
 	auth->passkey = passkey;
 
-	if (confirm_hint)
+	if (confirm_hint) {
+		if (device->bonding != NULL) {
+			/* We know the client has indicated the intent to pair
+			 * with the peer device, so we can auto-accept.
+			 */
+			btd_adapter_confirm_reply(device->adapter,
+						  &device->bdaddr,
+						  type, TRUE);
+			return 0;
+		}
+
 		err = agent_request_authorization(auth->agent, device,
 						confirm_cb, auth, NULL);
-	else
+	} else {
 		err = agent_request_confirmation(auth->agent, device, passkey,
 						confirm_cb, auth, NULL);
+	}
 
 	if (err < 0) {
 		if (err == -EINPROGRESS) {

