#########################################################################
Date: Feb 13, 2025 16:19 PM

Subject: Removed the unsupport BT device from the Bluez once we get the
Modalias of the device

Source: Sky
Upstream-Status: Pending
Signed-off-by: Ananth Marimuthu <ananth_marimuthu2@comcast.com>
#########################################################################

Index: bluez-5.48/src/device.c
===================================================================
--- bluez-5.48.orig/src/device.c
+++ bluez-5.48/src/device.c
@@ -99,6 +99,7 @@
 #define GATT_SND_SVC_UUID_STR  "2801"
 #define GATT_INCLUDE_UUID_STR "2802"
 #define GATT_CHARAC_UUID_STR "2803"
+#define XBOX_GEN4_UNSUPPORT_MODALIAS "usb:v045Ep0B13d0509"
 
 static DBusConnection *dbus_conn = NULL;
 static unsigned service_state_cb_id;
@@ -6431,6 +6432,19 @@ void btd_device_set_pnpid(struct btd_dev
 						DEVICE_INTERFACE, "Modalias");
 
 	store_device_info(device);
+
+#ifdef BT_UNSUPPORTED_GAMEPAD_ENABLED
+    if (g_strcmp0(device->modalias,XBOX_GEN4_UNSUPPORT_MODALIAS) == 0) {
+        DBG("Removed the unsupported device from the bluz %s", device->modalias);
+        btd_device_set_temporary(device, true);
+        if (!btd_device_is_connected(device)) {
+            DBG("btd_adapter_remove_device called here to remove device");
+            btd_adapter_remove_device(device->adapter, device);
+            return;
+        }
+        device_request_disconnect(device, NULL);
+    }
+#endif
 }
 
 static void service_state_changed(struct btd_service *service,
