Date: Nov 15, 2023
From: Srikanth Simgamsetti <srikanth_simgamsetti@comcast.com>
Source: COMCAST
Subject: Unable to reconnec to the paired device due to heavy cache list
Signed-off-by: Srikanth Simgamsetti <srikanth_simgamsetti@comcast.com>
Index: bluez-5.48/src/device.c
===================================================================
--- bluez-5.48.orig/src/device.c
+++ bluez-5.48/src/device.c
@@ -4132,6 +4132,7 @@ static void device_remove_stored(struct
 
 	g_free(data);
 	g_key_file_free(key_file);
+        remove(filename);
 }
 
 void device_remove(struct btd_device *device, gboolean remove_stored)
