Date: Mar 02 2023
From: schall488 <shivabhaskar_challa@comcast.com>
Subject: add missing patch of ping mdev support
Source: COMCAST
Signed-off-by: schall488 <shivabhaskar_challa@comcast.com>
Index: busybox-1.36.1/networking/ping.c
===================================================================
--- busybox-1.36.1.orig/networking/ping.c
+++ busybox-1.36.1/networking/ping.c
@@ -429,6 +429,7 @@ struct globals {
 	uint8_t pattern;
 	unsigned tmin, tmax; /* in us */
 	unsigned long long tsum; /* in us, sum of all times */
+	unsigned long long tsum2;
 	unsigned cur_us; /* low word only, we don't need more */
 	unsigned deadline_us;
 	unsigned interval_us;
@@ -458,6 +459,7 @@ struct globals {
 #define tmin         (G.tmin        )
 #define tmax         (G.tmax        )
 #define tsum         (G.tsum        )
+#define tsum2        (G.tsum2       )
 #define timeout      (G.timeout     )
 #define hostname     (G.hostname    )
 #define dotted       (G.dotted      )
@@ -500,10 +502,14 @@ static void print_stats_and_exit(int jun
 	printf("%lu%% packet loss\n", ul);
 	if (tmin != UINT_MAX) {
 		unsigned tavg = tsum / (nrecv + G.nrepeats);
-		printf("round-trip min/avg/max = %u.%03u/%u.%03u/%u.%03u ms\n",
-			tmin / 1000, tmin % 1000,
-			tavg / 1000, tavg % 1000,
-			tmax / 1000, tmax % 1000);
+                unsigned tavg2 = tsum2 / (nrecv + G.nrepeats);
+                unsigned long tmdev = sqrt(tavg2 - tavg * tavg );
+		printf("round-trip min/avg/max/mdev = %u.%03u/%u.%03u/%u.%03u/%u.%03u ms\n",
+				tmin / 1000, tmin % 1000,
+				tavg / 1000, tavg % 1000,
+				tmax / 1000, tmax % 1000,
+				tmdev / 1000, tmdev % 1000);
+
 	}
 	/* if condition is true, exit with 1 -- 'failure' */
 	exit(nrecv == 0 || (G.deadline_us && nrecv < pingcount));
@@ -661,6 +667,7 @@ static void unpack_tail(int sz, uint32_t
 		/* (doesn't hurt 32-bit real-world anyway) */
 		triptime = (int32_t) ((uint32_t)monotonic_us() - *tp);
 		tsum += triptime;
+		tsum2 += triptime * triptime;
 		if (triptime < tmin)
 			tmin = triptime;
 		if (triptime > tmax)
