--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -887,10 +887,6 @@ static int wiphy_info_handler(struct nl_msg *msg, void *arg)
         capa->flags |= WPA_DRIVER_FLAGS_OFFCHANNEL_TX;
     }
 
-	if (tb[NL80211_ATTR_ROAM_SUPPORT]) {
-		wpa_printf(MSG_DEBUG, "nl80211: Using driver-based roaming");
-		capa->flags |= WPA_DRIVER_FLAGS_BSS_SELECTION;
-	}
 
     wiphy_info_max_roc(capa,
                tb[NL80211_ATTR_MAX_REMAIN_ON_CHANNEL_DURATION]);