Index: NetworkManager-1.43.7/src/core/nm-connectivity.c
===================================================================
--- NetworkManager-1.43.7.orig/src/core/nm-connectivity.c
+++ NetworkManager-1.43.7/src/core/nm-connectivity.c
@@ -374,6 +374,11 @@ _con_curl_check_connectivity(CURLM *mhan
                                         NM_CONNECTIVITY_FULL,
                                         "no content, as expected",
                                         NULL);
+		_LOGD("Device connected to the internet for the first time");
+		g_bus_own_name(G_BUS_TYPE_SYSTEM,
+				"org.freedesktop.NetworkManager.connectivity",
+				G_BUS_NAME_OWNER_FLAGS_NONE,
+				NULL, NULL, NULL, NULL, NULL);
                 continue;
             }
 
@@ -383,6 +388,11 @@ _con_curl_check_connectivity(CURLM *mhan
                                         NM_CONNECTIVITY_FULL,
                                         "empty response, as expected",
                                         NULL);
+		_LOGD("Device connected to the internet for the first time");
+		g_bus_own_name(G_BUS_TYPE_SYSTEM,
+				"org.freedesktop.NetworkManager.connectivity",
+				G_BUS_NAME_OWNER_FLAGS_NONE,
+				NULL, NULL, NULL, NULL, NULL);
                 continue;
             }
         }
