Date: Jun 05 2025
From: Aravindan NC <nc.aravindan@gmail.com>
Subject: Add NetworkManager component in RDKE
Source: RDKCENTRAL
Signed-off-by: Aravindan NC <nc.aravindan@gmail.com>
Index: NetworkManager-1.43.7/data/NetworkManager.service.in
===================================================================
--- NetworkManager-1.43.7.orig/data/NetworkManager.service.in
+++ NetworkManager-1.43.7/data/NetworkManager.service.in
@@ -3,7 +3,7 @@ Description=Network Manager
 Documentation=man:NetworkManager(8)
 Wants=network.target
 After=network-pre.target dbus.service
-Before=network.target @DISTRO_NETWORK_SERVICE@
+Before=network.target @DISTRO_NETWORK_SERVICE@ wpa_supplicant.service
 BindsTo=dbus.service

 [Service]
@@ -11,6 +11,8 @@ Type=dbus
 BusName=org.freedesktop.NetworkManager
 ExecReload=/usr/bin/busctl call org.freedesktop.NetworkManager /org/freedesktop/NetworkManager org.freedesktop.NetworkManager Reload u 0
 #ExecReload=/bin/kill -HUP $MAINPID
+ExecStartPre= /bin/sh -c 'if [ ! -d /opt/NetworkManager/system-connections ];then mkdir -p /opt/NetworkManager/system-connections; fi'
+ExecStartPre= /bin/sh -c 'if [ ! -d /opt/secure/NetworkManager/system-connections ];then mkdir -p /opt/secure/NetworkManager/system-connections; fi'
 ExecStart=@sbindir@/NetworkManager --no-daemon
 Restart=on-failure
 # NM doesn't want systemd to kill its children for it
@@ -37,4 +38,4 @@ Also=NetworkManager-dispatcher.service
 # is enabled. NetworkManager-wait-online.service has
 # WantedBy=network-online.target, so enabling it only has an effect if
 # network-online.target itself is enabled or pulled in by some other unit.
-Also=NetworkManager-wait-online.service
+#Also=NetworkManager-wait-online.service
Index: NetworkManager-1.43.7/tools/meson-post-install.sh
===================================================================
--- NetworkManager-1.43.7.orig/tools/meson-post-install.sh
+++ NetworkManager-1.43.7/tools/meson-post-install.sh
@@ -21,7 +21,6 @@ if [ -x "${DESTDIR}${nm_bindir}/nmtui" ]
 fi

 for dir in "${nm_pkgconfdir}/conf.d" \
-           "${nm_pkgconfdir}/system-connections" \
            "${nm_pkgconfdir}/dispatcher.d/no-wait.d" \
            "${nm_pkgconfdir}/dispatcher.d/pre-down.d" \
            "${nm_pkgconfdir}/dispatcher.d/pre-up.d" \
