Index: NetworkManager-1.43.7/src/core/nm-policy.c
===================================================================
--- NetworkManager-1.43.7.orig/src/core/nm-policy.c
+++ NetworkManager-1.43.7/src/core/nm-policy.c
@@ -2092,7 +2092,6 @@ device_state_changed(NMDevice
 
             switch (nm_device_state_reason_check(reason)) {
             case NM_DEVICE_STATE_REASON_USER_REQUESTED:
-                blocked_reason = NM_SETTINGS_AUTOCONNECT_BLOCKED_REASON_USER_REQUEST;
                 break;
             case NM_DEVICE_STATE_REASON_DEPENDENCY_FAILED:
                 blocked_reason = NM_SETTINGS_AUTOCONNECT_BLOCKED_REASON_FAILED;
