Index: NetworkManager-1.43.7/src/core/nm-policy.c
===================================================================
--- NetworkManager-1.43.7.orig/src/core/nm-policy.c
+++ NetworkManager-1.43.7/src/core/nm-policy.c
@@ -2873,16 +2873,19 @@ dispose(GObject *object)
 {
     NMPolicy        *self = NM_POLICY(object);
     NMPolicyPrivate *priv = NM_POLICY_GET_PRIVATE(self);
-
-    nm_assert(!c_list_is_empty(&priv->policy_auto_activate_lst_head));
-    nm_assert(g_hash_table_size(priv->devices) == 0);
-
+    GHashTableIter   h_iter;
+    NMDevice        *device;
+
+
     nm_clear_g_object(&priv->default_ac4);
     nm_clear_g_object(&priv->default_ac6);
     nm_clear_g_object(&priv->activating_ac4);
     nm_clear_g_object(&priv->activating_ac6);
     nm_clear_pointer(&priv->pending_active_connections, g_hash_table_unref);

+    while ((device = c_list_first_entry(&priv->policy_auto_activate_lst_head, NMDevice, policy_auto_activate_lst)))
+        _auto_activate_device_clear(self, device, FALSE);
+
     g_slist_free_full(priv->pending_secondaries, (GDestroyNotify) pending_secondary_data_free);
     priv->pending_secondaries = NULL;

@@ -2901,6 +2906,12 @@ dispose(GObject *object)
         g_clear_object(&priv->dns_manager);
     }

+    g_hash_table_iter_init(&h_iter, priv->devices);
+    while (g_hash_table_iter_next(&h_iter, (gpointer *) &device, NULL)) {
+        g_hash_table_iter_remove(&h_iter);
+        devices_list_unregister(self, device);
+    }
+
     /* The manager should have disposed of ActiveConnections already, which
      * will have called active_connection_removed() and thus we don't need
      * to clean anything up.  Assert that this is TRUE.
