Index: NetworkManager-1.43.7/src/nmcli/common.c
===================================================================
--- NetworkManager-1.43.7.orig/src/nmcli/common.c
+++ NetworkManager-1.43.7/src/nmcli/common.c
@@ -930,7 +930,7 @@ read_again:

     /* If Ctrl-C was detected, complete the line */
     if (nmc_seen_sigint()) {
-        rl_echo_signal_char(SIGINT);
+        //rl_echo_signal_char(SIGINT);
         if (!rl_got_line) {
             rl_stuff_char('\n');
             rl_callback_read_char();
