Date: Nov 15, 2023
From: smuthu545 <Sindhuja_Muthukrishnan@comcast.com>
Source: COMCAST
Subject: Add thread info in heaptrack data
Signed-off-by: smuthu545 <Sindhuja_Muthukrishnan@comcast.com>
Index: git/src/arm/Gstep.c
===================================================================
--- git.orig/src/arm/Gstep.c
+++ git/src/arm/Gstep.c
@@ -107,8 +107,10 @@ unw_step (unw_cursor_t *cursor)
       ret = arm_exidx_step (c);
       Debug(1, "arm_exidx_step()=%d\n", ret);
       if (ret > 0)
+	      has_stopunwind = 1;
         return 1;
       if (ret == 0)
+	      has_stopunwind = 1;
         return ret;
       if (ret == -UNW_ESTOPUNWIND)
         has_stopunwind = 1;
@@ -125,16 +127,18 @@ unw_step (unw_cursor_t *cursor)
       Debug(1, "dwarf_step()=%d\n", ret);

       if (likely (ret > 0))
+	      has_stopunwind = 1;
         return 1;

       if (ret < 0 && ret != -UNW_ENOINFO)
         {
           Debug (2, "returning %d\n", ret);
+	  has_stopunwind = 1;
           return ret;
         }
     }
 #endif /* CONFIG_DEBUG_FRAME */
-
+has_stopunwind = 1;
   // Before trying the fallback, if any unwind info tell us to stop, do that.
   if (has_stopunwind)
     return -UNW_ESTOPUNWIND;
