Date: Jul 24, 2024
From: Mathew, Simi <Simi_Mathew@comcast.com>
Source: COMCAST
Subject: Moving between libsoup2 and libsoup3 will break
cookie jar db due to schema change.
Signed-off-by: Mathew, Simi <Simi_Mathew@comcast.com>
Index: libsoup-2.68.4/libsoup/soup-cookie-jar-db.c
===================================================================
--- libsoup-2.68.4.orig/libsoup/soup-cookie-jar-db.c
+++ libsoup-2.68.4/libsoup/soup-cookie-jar-db.c
@@ -132,7 +132,7 @@ soup_cookie_jar_db_new (const char *file
 
 #define QUERY_ALL "SELECT id, name, value, host, path, expiry, lastAccessed, isSecure, isHttpOnly FROM moz_cookies;"
 #define CREATE_TABLE "CREATE TABLE moz_cookies (id INTEGER PRIMARY KEY, name TEXT, value TEXT, host TEXT, path TEXT,expiry INTEGER, lastAccessed INTEGER, isSecure INTEGER, isHttpOnly INTEGER)"
-#define QUERY_INSERT "INSERT INTO moz_cookies VALUES(NULL, %Q, %Q, %Q, %Q, %d, NULL, %d, %d);"
+#define QUERY_INSERT "INSERT INTO moz_cookies (id, name, value, host, path,expiry, lastAccessed, isSecure, isHttpOnly) VALUES(NULL, %Q, %Q, %Q, %Q, %d, NULL, %d, %d);"
 #define QUERY_DELETE "DELETE FROM moz_cookies WHERE name=%Q AND host=%Q;"
 
 enum {

