Date: Oct 3, 2023
From: Srikanth Simgamsetti <srikanth_simgamsetti@comcast.com>
Source: COMCAST
Subject: Fix for Bluetoothd and BtrMgrBus process crash
Signed-off-by: Srikanth Simgamsetti <srikanth_simgamsetti@comcast.com>
Index: bluez-5.54/src/agent.c
===================================================================
--- bluez-5.54.orig/src/agent.c
+++ bluez-5.54/src/agent.c
@@ -218,7 +218,7 @@ void agent_unref(struct agent *agent)
 	if (agent->ref > 0)
 		return;
 
-	if (agent->request) {
+	if (agent->request && agent->request->cb) {
 		DBusError err;
 		agent_pincode_cb pincode_cb;
 		agent_passkey_cb passkey_cb;
@@ -252,7 +252,9 @@ void agent_unref(struct agent *agent)
 	}
 
 	g_free(agent->owner);
+        agent->owner = NULL;
 	g_free(agent->path);
+        agent->path = NULL;
 
 	g_free(agent);
 }
